src/Proxy/ProxyFactory.php

```diff
@@ -0,0 +1,142 @@
+<?php
+
+namespace GraphAware\Neo4j\OGM\Proxy;
+
+use GraphAware\Common\Type\Node;
+use GraphAware\Neo4j\OGM\EntityManager;
+use GraphAware\Neo4j\OGM\Metadata\NodeEntityMetadata;
+use GraphAware\Neo4j\OGM\Metadata\RelationshipMetadata;
+
+class ProxyFactory
+{
+    protected $em;
+
+    protected $classMetadata;
+
+    protected $proxyDir;
+
+    public function __construct(EntityManager $em, NodeEntityMetadata $classMetadata)
+    {
+        $this->em = $em;
+        $this->classMetadata = $classMetadata;
+        $this->proxyDir = $em->getProxyDirectory();
+    }
+
+    public function fromNode(Node $node)
+    {
+        $object = $this->createProxy();
+        $object->__setNode($node);
+        $initializers = [];
+        foreach ($this->classMetadata->getSimpleRelationships() as $relationship) {
+            $initializer = $this->getInitializerFor($relationship);
+            $initializers[$relationship->getPropertyName()] = $initializer;
+        }
+        $object->__setInitializers($initializers);
+
+        return $object;
+    }
+
+    private function getInitializerFor(RelationshipMetadata $relationship) {
+        if (!$relationship->isCollection()) {
+            $initializer = new SingleNodeInitializer($this->em, $relationship, $this->em->getClassMetadata($relationship->getTargetEntity()));
+        }
+
+        return $initializer;
+    }
+
+    protected function createProxy()
+    {
+        $class = $this->classMetadata->getClassName();
+        $proxyClass = $this->getProxyClass();
+        $proxyFile = $this->proxyDir.'/'.$proxyClass.'.php';
+        $methodProxies = $this->getMethodProxies();
+
+        $content = <<<PROXY
+<?php
+
+use GraphAware\\Neo4j\\OGM\\Proxy\\EntityProxy;
+
+class $proxyClass extends $class implements EntityProxy
+{
+    private \$em;
+    private \$initialized = [];
+    private \$intializers = [];
+    private \$node;
+    
+    public function __setNode(\$node)
+    {
+        \$this->node = \$node;
+    }
+    
+    public function __setInitializers(array \$initializers)
+    {
+        \$this->initializers = \$initializers;
+    }
+    
+    public function __initializeProperty(\$propertyName)
+    {
+        \$value = \$this->initializers[\$propertyName]->initialize(\$this->node);
+        \$this->\$propertyName = \$value;
+    }
+    
+    $methodProxies
+}
+
+PROXY;
+
+        $this->checkProxyDirectory();
+        file_put_contents($proxyFile, $content);
+
+        if (!class_exists($proxyClass)) {
+            require $proxyFile;
+        }
+
+        return $this->newProxyInstance($proxyClass);
+
+    }
+
+    protected function getMethodProxies()
+    {
+        $proxies = '';
+        foreach ($this->classMetadata->getRelationships() as $relationship) {
```

Comment:
In Doctrine, when an entity has relationships, a proxy is created, like here. The difference is that in Doctrine, they pre-fetch the `row id` of the relations, this is why you can have proxies that don't have their properties hydrated yet and thus should be initialized.

I don't want to take this approach here, if we traverse relationships we return a complete version of the other node, I have two reasons for it : 

* Returning a node object or just the id doesn't make real differences on the driver level (actually with bolt it is faster to return node objects than node ids for eg).

* Some might argue that this will have lot of difference when nodes contains lot of properties or properties with huge property values -> then it is their usage of neo4j that is wrong, you tend generally to keep your nodes clean and direct, the real power of a graph is traversing, use elasticsearch for indexing text etc.. This is something I'll be happy to review once some come with performance problem that is not a neo4j mis-usage.

Also, if I can first make it "work" and be stable with some features that makes us as devs happy and proficient, then optimizations will come. There is still some work to do before reaching 1.0RC1. 

But thanks for the reviews and the encouragements.